### PR TITLE
Implement support for redis clusters

### PR DESCRIPTION
**Pull request by evanpurkhiser**
This adds support for configuring and using true redis clusters.

This adds a new module variable `sentry.utils.redis.redis_clusters` which is an instance of the `ClusterManager` that supports creation and retrieval of `StrictRedisCluster` clients. This deviates from the usage of `ClusterManager` that manages rb `Cluster` objects, thus the separate module variable from `clusters`.

Nothing actually uses this, but will link to the PR for the similarity engine using it here once it's ready.

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes to build requirements</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by EvanPurkhiser** [comment]
adding some test cases for this.

File: setup.py
Comment (tkaemming) [inline setup.py]: Are we committing to this being a universally required dependency out of the gate (e.g. for on premise?)

I guess there's probably no harm in it … ?

File: src/sentry/data/config/config.yml.default
Comment (tkaemming) [inline src/sentry/data/config/config.yml.default]: "True" Redis Cluster is a weird wording — I'd clarify the distinction between rb and Redis "capital-C" Cluster here.

File: src/sentry/data/config/config.yml.default
Comment (tkaemming) [inline src/sentry/data/config/config.yml.default]: `s/inteded/intended/`

File: src/sentry/utils/redis.py
Comment (tkaemming) [inline src/sentry/utils/redis.py]: Isn't this the same as `config.get('hosts').values()`?

I'm not familiar with the options for Redis Cluster, but it seems like we shouldn't try to make the configuration parameters here superficially similar just for consistency (for example, this configuration should probably just include a `nodes` list rather than a `hosts` map that we transform into the node list) unless there was a wider plan for this to simplify the upgrade path or something?

**Review by tkaemming — COMMENTED** [review]
Code itself looks fine, added a few questions and nitpicky stuff inline. I didn't approve yet since I'm not sure if we're ready to use it yet, let me know

File: src/sentry/utils/redis.py
Comment (EvanPurkhiser) [inline src/sentry/utils/redis.py]: > Isn't this the same as config.get('hosts').values()?

Yeah it totally is, surprised I missed that.

--

Are you saying in the `redis.clusters` configuration, it should just be formatted differently than how we're formatting the rb hosts?

Comment (EvanPurkhiser) [inline src/sentry/utils/redis.py]: https://github.com/getsentry/sentry/pull/5883/commits/7acf6b6319bb817fe8cb43489a6e989d5d5ee8ba

File: src/sentry/data/config/config.yml.default
Comment (EvanPurkhiser) [inline src/sentry/data/config/config.yml.default]: https://github.com/getsentry/sentry/pull/5883/commits/cd3613b1d5893a9d1934d5aeab0b61c68969923b

File: src/sentry/data/config/config.yml.default
Comment (EvanPurkhiser) [inline src/sentry/data/config/config.yml.default]: https://github.com/getsentry/sentry/pull/5883/commits/cd3613b1d5893a9d1934d5aeab0b61c68969923b

**Comment by EvanPurkhiser** [comment]
No need to merge right away, but I think this should be in good shape now.

File: src/sentry/utils/redis.py
Comment (tkaemming) [inline src/sentry/utils/redis.py]: > Are you saying in the `redis.clusters` configuration, it should just be formatted differently than how we're formatting the rb hosts?

Yeah. I think the rb configuration is the way it is because the previous clustering tool (nydus) required this format and I'm not sure there is a compelling reason to continue using the "map with monotonically increasing integer keys that starts at 0 without gaps" format when an array/list type will work just fine.

Comment (EvanPurkhiser) [inline src/sentry/utils/redis.py]: Yeah definitely agree with this. I'll make another PR adds support for a list, but continues to support a map

Comment (EvanPurkhiser) [inline src/sentry/utils/redis.py]: 20fcd94c0f87b29062458699be45be7e77785611 Just added to here.

**Comment by EvanPurkhiser** [comment]
Rebased on top of master and added soft fail on connection startup as discussed with @mattrobenolt internally.

File: setup.py
Comment (mattrobenolt) [reply setup.py]: I don't see a problem with it. But I would recommend capping the max version. Too many problems with dependency versions lately.

File: src/sentry/utils/redis.py
Comment (mattrobenolt) [inline src/sentry/utils/redis.py]: It's not obvious to me, but will this attempt to reconnect? Or if it fails to connect at startup, is it just dead for the whole life of the process?

Comment (EvanPurkhiser) [reply src/sentry/utils/redis.py]: Right now I'm leaving that up to the caller, so if it makes another attempt later it will try to construct the client again.

File: setup.py
Comment (EvanPurkhiser) [reply setup.py]: d215848e885ae7b5243a46206f043cb73929f238

**Comment by EvanPurkhiser** [comment]
Rebased onto master

@mattrobenolt: Wanted to get your final 👍  on this.

 * When asking for a cluster, it will now raise a `KeyError` if it's unable to connect to the cluster. Unlike RB redis cluster does not lazily connect. I thought about wrapping the `RedisClusterClient` so that it wouldn't attempt to connect until it's first used, but I dislike the semantics there.

 * As per @tkaemming's comment on supporting list style hosts, you can now configure the `redis.clusters` dict with either a list of hosts, or map of hosts. I wanted to support both styles for both types to keep things consistent.

**Review by mattrobenolt — APPROVED** [review]
As long as @tkaemming is cool with it, I'm cool with it. 😎

**Comment by tkaemming** [comment]
> As long as @tkaemming is cool with it, I'm cool with it. 😎 

Gopher it

**Comment by EvanPurkhiser** [comment]
Cool, waiting for the final 👍  on https://github.com/getsentry/sentry/pull/5891 and we're 🥇

**Pull request merged**
